stage1/diagexec/diagexec.c

```diff
@@ -146,16 +146,48 @@ static void diag(const char *exe)
 	diag(itrp);
 }
 
-/* Read environment from env and make it our own. */
-/* The environment file must exist, may be empty, and is expected to be of the format:
+/* Append env variables listed in keep_env to env_file if they're present in
+ * current environment */
+void append_env(const char *env_file, const char **keep_env)
+{
+	FILE		*f;
+	const char	**p;
+	char		*v;
+	char		nul = '\0';
+
+	pexit_if((f = fopen(env_file, "a")) == NULL,
+		"Unable to fopen \"%s\"", env_file);
+
+	p = keep_env;
+	while (*p) {
+		v = getenv(*p);
+		if (v) {
+			fprintf(f, "%s=%s", *p, v);
```

Comment:
Oh yeah, and `fprintf` return value should be compared against `strlen(*p) + strlen(v) + 2`.